Reduce Promise forking

btbonval opened the pull request:
Addresses Issue #1  

Some notes:

* `App.then()` returns `this`, which is thenable. Abstracts away the need to mess with `this.handlePromise`
* added App.catch()
* created `HandleSynchronisedSafeChild` which constructs with a `HandleSynchronisedSafeObject` parent
* `MutableData` is a Child, so `MutableData.then` and `MutableData.catch` operate on the parent's `handlePromise` to avoid forking and keep Promise chain order obvious.
* added `App.free()`
* track App's handle using `this.handleLocal`: if `App.free()` is called after `App.catch()`, there is no way the handle is going to get returned from the previous Promise.
* added `safeApp.refreshContainersPermissions()` to `App.authoriseAndConnect()`; I don't know what it does, but it is used at that time in every working example I've seen.
* added `App.quickRandomPublic(obj)` which creates and returns a random, public MutableData containing key/value pairs from `obj`.
* added `MutableData.getString(key)` which gets the value for the given `key` and converts that value to a string.
* modified `MutableData.free()`. It isn't obvious to me that `mdHandle` will necessarily get passed in, and it is saved into `MutableData` (at least it is in `quickRandomPublic`). It'll free whatever source it can get so long as everything looks consistent.

btbonval committed:
streamlining MutableData to a single, non-forked Promise chain

btbonval committed:
Merge commit '5b65a527b5808e3e82b7611b16eaa5c9b2f9c082'

Conflicts:
	safe_api.js

loziniak commented on safe_api.js:
```diff
@@ -78,10 +133,41 @@ window.safeAPI = {};
-				this.handlePromise = null;
+			this.then((mdHandle) => {
+				// Nothing to do if no handles are provided
+				if ((mdHandle === undefined) && (this.mdHandle === undefined)) {
```
where do you set this.mdHandle for the mutable data object?

btbonval replied:
For many of these that I wasn't personally using, I just left the mdHandle as-is and only changed the `then()` calls. I was assuming that you already had your mdHandle passed properly in any code you already wrote.

loziniak commented on safe_api.js:
```diff
@@ -78,10 +133,41 @@ window.safeAPI = {};
-			this.handlePromise.then((mdHandle) => {
-				safeMutableData.free(mdHandle);
-				this.handlePromise = null;
+			this.then((mdHandle) => {
```
You will get `appHandle` as `mdHandle` here, because `this.then` calls `this.parent.then` ?

btbonval replied:
`mdHandle` depends entirely on whatever the last function returns. It might not even be an mdHandle. It might be null if the previous `thenable` called has a Promise that returns nothing.

That's why in the code I was using, I switched to `self.handle` in the code I was personally using and tested.

loziniak replied:
Thanks, but I think having a cleaner code is more important than breaking existing one :-) So break everything you want.

loziniak commented on safe_api.js:
```diff
@@ -78,10 +133,41 @@ window.safeAPI = {};
+				};
+				// only mdHandle is defined, clear it.
+				if ((mdHandle != undefined) && (this.mdHandle === undefined)) {
+					safeMutableData.free(mdHandle);
```
mdHandle will be appHandle in fact?

btbonval replied:
`this.mdHandle` will be the `mdHandle`, no question.

You put `mdHandle` in as a function call, which I left. For any code you wrote, you're assuming that the previous call before `free()` is returning the `mdHandle`. It might be the `mdHandle`, it might be the `appHandle`, it entirely depends on the return of the previous `thenable` called.

This is why I started storing `self.handle`: no confusion.

But I didn't want to break any code you already wrote, so I supported passing in the old value.

btbonval commented:
As a side note, I've further refactored this in my branch of your code.

It is still untested, but you can see where I'm at here:

https://github.com/btbonval/maidsafe-test/blob/untested/safe_api.js

I pulled `free()` into the parent classes. I rewrote _all_ the functions instead of just the ones I was using. I'm trying to avoid the sort of code where I do this:

```
var blah = new Class();
this.then((handle) => {
  blah.handle = handle;
}
return blah;
```

Because in the above code, `blah` is constructed and returned _before_ `blah.handle` is written. In the example above, it is a bit annoying that setting the handle is required for the object to work, but setting the handle is not part of the constructor. In my branch, I'm attempting to capture the Promise returning that handle directly into the `new Class()` constructor:

```
		newMutation() {
			var newMut;
			this.then(_ => safeMutableData.newMutation(this.handle))
			.then((mHandle) => {
				newMut = new Mutation(mHandle, this);
			});
			return newMut;
		}
```

Testing isn't complete yet. I don't think this will work because newMut will return as null until the Promise finishes. In fact, I'm quite sure my branch won't work as it is presently written. I've already rewritten it on my local system and I am testing it right now. If I get it working, I'll submit another PR with working code that is a lot cleaner _and_ consistent throughout.
